src/software/new_geom/cubic_bezier_spline2d_test.cpp

```diff
@@ -0,0 +1,235 @@
+        const Point knot             = test_spline_1.getValueAt(knot_input_value);
+        const Point just_after_knot  = test_spline_1.getValueAt(knot_input_value - 1e-9);
+
+        const Vector first_derivative_just_before_knot = knot - just_before_knot;
+        const Vector first_derivative_just_after_knot  = just_after_knot - knot;
+
+        // C1 continuity means we should have the same (approximately) the same first
+        // derivative before and after the knot
+        const Vector first_derivative_change =
+            first_derivative_just_after_knot - first_derivative_just_before_knot;
+        EXPECT_NEAR(first_derivative_change.x(), 0, 1e-7);
+        EXPECT_NEAR(first_derivative_change.y(), 0, 1e-7);
+
+        const Point just_before_just_before_knot =
+            test_spline_1.getValueAt(knot_input_value - 2 * 1e-9);
+        const Point just_after_just_after_knot =
+            test_spline_1.getValueAt(knot_input_value - 2 * 1e-9);
+
+        const Vector first_derivative_just_before_just_before_knot =
+            just_before_knot - just_before_just_before_knot;
+        const Vector first_derivative_just_after_just_after =
+            just_after_just_after_knot - just_after_knot;
+
+        const Vector second_derivative_just_before_knot =
+            first_derivative_just_before_knot -
+            first_derivative_just_before_just_before_knot;
+        const Vector second_derivative_just_after_knot =
+            first_derivative_just_after_just_after - first_derivative_just_after_knot;
+
+        // C2 continuity means we should have the same (approximately) the same second
+        // derivative before and after the knot
+        const Vector second_derivative_change =
+            second_derivative_just_after_knot - second_derivative_just_before_knot;
+
+        EXPECT_NEAR(second_derivative_change.x(), 0, 1e-9);
+        EXPECT_NEAR(second_derivative_change.y(), 0, 1e-9);
+    }
+}
+
+TEST_F(CubicBezierSplineTest, getControlPoints_no_intermediate_knots)
+{
+    CubicBezierSpline2d spline(Point(0, 0), Vector(0, 1), Point(2, 0), Vector(0, 1), {});
+
+    std::vector<Point> expected_control_points = {Point(0, 0), Point(0, 1), Point(2, 1),
+                                                  Point(2, 0)};
+    EXPECT_EQ(expected_control_points, spline.getControlPoints());
+}
+
+TEST_F(CubicBezierSplineTest, getControlPoints_single_intermediate_knot)
+{
+    // This spline is effectively the first one shown on this page:
+    // https://www.ibiblio.org/e-notes/Splines/b-int.html
+    Point p0(0, 0);
+    Point p1(1, 0);
+    Point p2(1.5, -1.5);
+    Vector d0(0.4, 0.4);
+    Vector d2(0.1, -0.1);
+    Vector d1 = (p2 - p0 - d2 - d0) / 4.0;
+
+    // We negate d2 here so that we can look at this like it's represented on the
+    // website linked above, since our convention is that the final vector is pointing
+    // back "towards the start" of the spline, rather then "off then end" of it
```

Comment:
Same comment about the end point not pointing towards the *start* of the spline

Reply:
changed to `back along the body of the spline`